File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    if (!words.includes(target)) return 0
    const visited = new Array(words.length).fill(false)
    const q = []
    q.push({ word: begin, depth: 0})
    
    while (q.length) {
        const {word, depth} = q.shift()
        if (word === target) return depth
        words.forEach((w, idx) => {
            if (!visited[idx] && verify(word, w)) {
                q.push({word: w, depth:depth+1})
                visited[idx] = 1 
                }
        })
    }
    
    
    return answer;
}
    
function verify(word, w) {
    let diffCnt = 0
    for (let i = 0; i < word.length; i++) {
        if (word[i] !== w[i]) diffCnt++
    }
    return diffCnt === 1 ? true : false
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    return bfs(words, begin, target)

def bfs(words, begin, target):
    if target not in words:
        return 0
    
    q = deque()
    for word in words:
        if check(begin, word):
            q.append([word, 1])
    
    while q:
        word, cnt = q.popleft()
        if word == target:
            return cnt
        for w in words:
            if check(word, w):
                if w == target:
                    return cnt+1
                q.append([w, cnt+1])
    return 0
        
def check(word1, word2):
    cnt = 0
    for i in range(len(word1)):
        if word1[i] == word2[i]:
            cnt += 1
    if cnt == (len(word1) - 1):
        return True
    else:
        return False